ethcore/src/engines/authority_round/finality.rs

```diff
@@ -110,97 +104,127 @@ impl RollingFinality {
 	/// Fails if `signer` isn't a member of the active validator set.
 	/// Returns a list of all newly finalized headers.
 	// TODO: optimize with smallvec.
-	pub fn push_hash(&mut self, head: H256, signers: Vec<Address>) -> Result<Vec<H256>, UnknownValidator> {
+	pub fn push_hash(&mut self, head: H256, number: BlockNumber, signers: Vec<Address>)
+		-> Result<Vec<H256>, UnknownValidator>
+	{
 		for their_signer in signers.iter() {
 			if !self.signers.contains(their_signer) {
 				warn!(target: "finality",  "Unknown validator: {}", their_signer);
 				return Err(UnknownValidator)
 			}
 		}
 
-		for signer in signers.iter() {
-			*self.sign_count.entry(*signer).or_insert(0) += 1;
-		}
-
-		self.headers.push_back((head, signers));
+		self.add_signers(&signers);
+		self.headers.push_back((head, number, signers));
 
 		let mut newly_finalized = Vec::new();
 
-		while self.sign_count.len() * 2 > self.signers.len() {
-			let (hash, signers) = self.headers.pop_front()
+		while self.is_finalized() {
+			let (hash, _, signers) = self.headers.pop_front()
 				.expect("headers length always greater than sign count length; qed");
-
+			self.remove_signers(&signers);
 			newly_finalized.push(hash);
-
-			for signer in signers {
-				match self.sign_count.entry(signer) {
-					Entry::Occupied(mut entry) => {
-						// decrement count for this signer and purge on zero.
-						*entry.get_mut() -= 1;
-
-						if *entry.get() == 0 {
-							entry.remove();
-						}
-					}
-					Entry::Vacant(_) => panic!("all hashes in `header` should have entries in `sign_count` for their signers; qed"),
-				}
-			}
 		}
 
 		trace!(target: "finality", "{} Blocks finalized by {:?}: {:?}", newly_finalized.len(), head, newly_finalized);
 
 		self.last_pushed = Some(head);
 		Ok(newly_finalized)
 	}
+
+	/// Returns the first block for which a 2/3 quorum (instead of 1/2) is required.
+	pub fn quorum_2_3_transition(&self) -> BlockNumber {
+		self.quorum_2_3_transition
+	}
+
+	/// Returns whether the first entry in `self.headers` is finalized.
+	fn is_finalized(&self) -> bool {
+		match self.headers.front() {
+			None => false,
+			Some((_, number, _)) if *number < self.quorum_2_3_transition => {
+				self.sign_count.len() * 2 > self.signers.len()
+			}
+			Some((_, _, _)) => {
+				self.sign_count.len() * 3 > self.signers.len() * 2
+			}
+		}
+	}
+
+	/// Adds the signers to the sign count.
+	fn add_signers(&mut self, signers: &[Address]) {
+		for signer in signers {
+			*self.sign_count.entry(*signer).or_insert(0) += 1;
+		}
+	}
+
+	/// Removes the signers from the sign count.
+	fn remove_signers(&mut self, signers: &[Address]) {
+		for signer in signers {
+			match self.sign_count.entry(*signer) {
+				Entry::Occupied(mut entry) => {
+					// decrement count for this signer and purge on zero.
+					*entry.get_mut() -= 1;
```

Comment:
Maybe I'm paranoid but if we ever call `remove_signers()` with an `Occupied` entry that is `0`, then this will underflow and the signer will not be removed. Maybe you can do `if *entry.get() == 1` before this and remove it if it is (and decrement if it isn't)?